### DISCUSSION TITLE
docker K8s local - Ytt - kapp

### DISCUSSION THREAD

**Discussion by enkicoma** [k14s/ytt#293] [Q&A] [answered]
Hi Guys, I am facing an issue:
```
kapp deploy -a cf -f <(ytt -f config -f ./cf-install-values.yml -f ./config/remove-resource-requirements.yml)
ytt: Error: Overlaying data values (in following order: values/00-values.yml, values/10-images.yml, values/20-secrets-config-values.yml, values/30-experimental-values.yml, cf-install-values.yml):
  Document on line cf-install-values.yml:2:
    Map item (key 'kpack') on line cf-install-values.yml:338:
      Expected number of matched nodes to be 1, but was 0
Target cluster 'https://127.0.0.1:59508' (nodes: kind-control-plane)
```
tried before to run:
`kapp deploy -a cf -f <(ytt -f config -f ./cf-install-values.yml -f ./config-optional/remove-resource-requirements.yml -f ./config-optional/remove-ingressgateway-service.yml)` but looks like `remove-ingressgateway-service.yml` has been removed

cf-install-values.yml :
```
kpack:
  registry:
    hostname: https://index.docker.io/v1/
    repository: cloudfoundry/capi
    username: "%$$"
    password: "$$-$$-$$-$$-$$"
```
before I tried:
```
app_registry:
  hostname: https://index.docker.io/v1/
  repository: "**"
  username: "**"
  password: "***"
```
the same issue...  hope someone can guide me in the right direction, cheers!

**Comment by enkicoma** [comment, marked as answer]
I think I followed an old document:
here is the latest:
https://github.com/cloudfoundry/cf-for-k8s/blob/develop/docs/getting-started-tutorial.md

and it works

**Comment by joaopapereira** [comment]
Hey @enkicoma 
I assume that you are trying to use https://github.com/cloudfoundry/cf-for-k8s, is that correct?

After some digging around, this is what I found. The error message that you see there should read as follow

ytt tried to merge together all these data values files (values/00-values.yml, values/10-images.yml, values/20-secrets-config-values.yml, values/30-experimental-values.yml, cf-install-values.yml) but because kpack key is not present in values/00-values.yml ytt does not know how to interpret it.

In other words, the kpack key is never referenced in any of the values files, except for cf-install-values.yml, so it is not valid.

ytt Specific Solution:
when you want to add extra values to your configuration, like the kpack example, you need to create an overlay that looks somewhat like this:
```yaml
#@data/values
#@overlay/match-child-defaults missing_ok=True
---
kpack:
  registry:
    hostname: https://index.docker.io/v1/
    repository: cloudfoundry/capi
    username: "%$$"
    password: "$$-$$-$$-$$-$$"
```
The emphasis here is in the annotation `#@overlay/match-child-defaults missing_ok=True` that tells ytt that it is ok if there are child keys that didn't exist before.  For more information take a look at https://carvel.dev/ytt/docs/latest/lang-ref-ytt-overlay/#overlaymatch-child-defaults

cf-for-k8s Specific Solution:
Looks like that kpack relies on the `app_registry` configuration to setup the needed registry variables, so you just need to set the values you that part of the configuration.


Another thing that I noticed is that you are doing `ytt -f config -f ./cf-install-values.yml -f ./config/remove-resource-requirements.yml` this will try to execute twice the file `./config/remove-resource-requirements.yml` which will cause the following error:
```
ytt: Error: Overlaying (in following order: 2-fix-null-annotations.yml, capi/allow-prometheus-metrics-access.yml, capi/capi.yml, capi/staging-ns-label.yml, eirini/eirini.yml, eirini/enable-automount-service-account-token.yml, fix-db-startup-order.yml, istio/add-istio-injection.yml, istio/external-routing.yml, istio/ingressgateway-service-nodeport.yml, istio/istio-kapp-ordering.yml, istio/label-istio-ns.yml, istio/remove-hpas-and-scale-istiod.yml, istio/use-external-dns-for-wildcard.yml, istio/use-first-party-jwt-tokens.yml, kpack/kapp-order.yml, kpack/kpack.yml, metrics/allow-prometheus-metrics-access.yml, metrics/metrics-server-prefer-internal-kubelet-address.yml, minio/minio.yml, networking/network-policies.yaml, postgres/postgres.yml, prioritize-daemonsets.yml, remove-resource-requirements.yml, uaa/allow-prometheus-metrics-access.yml, uaa/enable-automount-service-account-token.yml, uaa/remove-resource-requirements.yml, uaa/uaa.yml, workloads-namespace.yml, z-kapp-versioned-creds.yml, remove-resource-requirements.yml):
  Document on line remove-resource-requirements.yml:36:
    Map item (key 'data') on line remove-resource-requirements.yml:37:
      Map item (key 'values') on line remove-resource-requirements.yml:39: overlay.apply:
        Map item (key 'global') on line remove-resource-requirements.yml:29:
          Map item (key 'proxy_init') on line remove-resource-requirements.yml:30:
            Map item (key 'resources') on line remove-resource-requirements.yml:32:
              Expected number of matched nodes to be 1, but was 0
```
you can remove the last -f

**Reply by enkicoma** [reply]
Yes, indeed I tried to use https://github.com/cloudfoundry/cf-for-k8s 

A very professional approach on explaining the issue, thank you ^_^
